### PR TITLE
fix(core): handle MiMo tool-result media

### LINKED ISSUE #4223: mimo-v2.5-pro API Error: 400 Param Incorrect
### What happened?

Image

mimo-v2.5-pro首次调用工具正常执行\
再次调用工具时报错400 Param Incorrect

根据网络搜索结果来看, 多半是reasoning_content字段的问题, 和DeepseekV4Pro一样

值得一提的是, 一周前使用都是正常的, 最近突然频繁报错
(多次Ctrl+Y重试, 有概率成功继续运行)

### What did you expect to happen?

Image

### Client information

Qwen Code: 0.15.11 (782403d71)
运行环境: Node.js v22.12.0 / npm 10.9.0
操作系统: win32 x64 (10.0.26200)
认证: API Key - openai
Base URL: https://jiuuij.de5.net/v1
模型: qwen3.6-max-preview-thinking
快速模型: gemini-3-flash-preview
会话 ID: 4a3c4d79-847c-4f73-a580-1759293a35a5
沙箱: no sandbox
代理: no proxy
内存使用: 233.2 MB

### Login information

API Config

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by DragonnZhang**
## Summary

- What changed: Added a MiMo OpenAI-compatible provider and a provider request-context override so MiMo requests preserve tool-call reasoning content and split media returned by tools out of `role: \"tool\"` messages.
- Why it changed: MiMo-compatible chat endpoints can reject follow-up requests with `400 Param Incorrect` when tool-call history omits `reasoning_content`, and strict chat-completions servers can also reject image/video/file content embedded directly in tool result messages.
- Reviewer focus: Please check the provider selection logic, the `splitToolMedia` default for MiMo, and the request-context override precedence.

## Validation

- Commands run:
  ```bash
  cd packages/core && npx vitest run src/core/openaiContentGenerator/provider/mimo.test.ts src/core/openaiContentGenerator/pipeline.test.ts
  cd packages/core && npm run typecheck
  npm run typecheck
  ```
- Prompts / inputs used: Reproduced from session `1f1a7c9d-f054-46fd-8458-1717cd611968`, where the user asked Qwen Code to read any repository image and `read_file` returned a PNG as inline media.
- Expected result: MiMo requests keep tool messages text-only when tool results include media, while still replaying `reasoning_content` for assistant tool-call turns.
- Observed result: The two focused test files passed with 50 tests, and `packages/core` typecheck passed.
- Quickest reviewer verification path: Run the two focused Vitest files above, then try a MiMo session that reads a PNG via `read_file` and verify the follow-up request no longer fails with `400 Param Incorrect`.
- Evidence (output, logs, screenshots, video, JSON, before/after, etc.): The failing local session showed `read_file` returning `inlineData` for `icon-orange.png`, followed immediately by `qwen-code.api_error status=400 model=mimo-v2.5 msg=400 Param Incorrect`.

## Scope / Risk

- Main risk or tradeoff: MiMo now defaults tool-result media splitting on, which changes the message shape from media inside the tool message to a text-only tool message followed by user media.
- Not covered / not validated: I did not run a live MiMo API call after the fix.
- Breaking changes / migration notes: None expected. Users can still explicitly set `splitToolMedia: false` to opt out.

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ⚠️  | ⚠️  |
| npx      | ✅  | ⚠️  | ⚠️  |
| Docker   | ⚠️  | N/A | ⚠️  |
| Podman   | ⚠️  | N/A | ⚠️  |
| Seatbelt | ⚠️  | N/A | N/A |

Testing matrix notes:

- Tested on macOS with focused `npx vitest` and `packages/core` `npm run typecheck`.
- Root `npm run typecheck` was attempted but is currently blocked by unrelated existing `packages/cli/src/serve/workspaceAgents.ts` and `workspaceMemory.ts` type errors.
- Windows, Linux, Docker, Podman, and Seatbelt were not run locally.

## Linked Issues / Bugs

Fixes #4223

### CODE REVIEW

File: packages/core/src/core/openaiContentGenerator/provider/mimo.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      return baseRequest;
+    }
+
+    return {
+      ...baseRequest,
+      messages: baseRequest.messages.map(ensureReasoningContentOnToolCalls),
+    };
+  }
+
+  getRequestContextOverrides(): { splitToolMedia?: boolean } {
```
Comment (wenshao) [inline packages/core/src/core/openaiContentGenerator/provider/mimo.ts]: **[Suggestion]** `getRequestContextOverrides()` uses an inline return type `{ splitToolMedia?: boolean }` instead of the `OpenAIRequestContextOverrides` type defined in `types.ts` (added in this same PR). If the interface gains new fields later, this implementation won't be flagged by the type checker.

```suggestion
  getRequestContextOverrides(): OpenAIRequestContextOverrides {
```

(Also add `import type { OpenAIRequestContextOverrides } from './types.js';` at the top.)

_— qwen-latest-series-invite-beta-v28 via Qwen Code /review_

File: packages/core/src/core/openaiContentGenerator/pipeline.test.ts [hunk 2]
```diff
@@ -216,6 +216,48 @@ describe('ContentGenerationPipeline', () => {
       );
     });
 
+    it('should apply provider request context overrides', async () => {
```
Comment (wenshao) [inline packages/core/src/core/openaiContentGenerator/pipeline.test.ts:219]: **[Suggestion]** This test verifies that a provider override of `{ splitToolMedia: true }` reaches the converter, but `contentGeneratorConfig.splitToolMedia` is `undefined` in the test setup — so the override trivially wins. The actual precedence chain (`providerOverrides.splitToolMedia ?? contentGeneratorConfig.splitToolMedia ?? false`) is never exercised with conflicting values.

Consider adding a test where `contentGeneratorConfig.splitToolMedia = true` and the provider returns `{ splitToolMedia: false }`, asserting the converter receives `splitToolMedia: false`. This would catch a future inversion of the precedence order.

_— qwen-latest-series-invite-beta-v28 via Qwen Code /review_

File: packages/core/src/core/openaiContentGenerator/provider/mimo.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  if (message.role !== 'assistant') {
+    return message;
+  }
+
+  const assistant = message as ExtendedChatCompletionAssistantMessageParam;
+  if (!assistant.tool_calls?.length) {
+    return message;
+  }
+
+  if (typeof assistant.reasoning_content === 'string') {
```
Comment (wenshao) [inline packages/core/src/core/openaiContentGenerator/provider/mimo.ts]: **[Suggestion]** `ensureReasoningContentOnToolCalls` diverges from the DeepSeek provider's equivalent function in two ways:

1. **Guard:** DeepSeek checks `typeof === 'string' && .length > 0` (re-injects when empty string), MiMo checks only `typeof === 'string'` (preserves empty string as-is).
2. **Scope:** DeepSeek applies to ALL assistant turns; MiMo only applies to turns with `tool_calls`.

The scope difference appears intentional, but the empty-string handling divergence is a maintenance hazard — two near-identical functions with subtly different logic. If someone fixes a bug in one copy, they may forget to port it to the other.

Consider extracting a shared helper into `provider/utils.ts` that accepts an `options` parameter to control these differences, or add a comment documenting why the divergence is intentional.

_— qwen-latest-series-invite-beta-v28 via Qwen Code /review_

File: packages/core/src/core/openaiContentGenerator/provider/mimo.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    };
+  }
+}
+
+function ensureReasoningContentOnToolCalls(
+  message: OpenAI.Chat.ChatCompletionMessageParam,
+): OpenAI.Chat.ChatCompletionMessageParam {
+  if (message.role !== 'assistant') {
+    return message;
+  }
```
Comment (wenshao) [inline packages/core/src/core/openaiContentGenerator/provider/mimo.ts]: **[Critical]** `reasoning_content` is only added when an assistant message has `tool_calls`, but the linked failure mode is the same thinking-mode contract as DeepSeek: replayed assistant turns must carry `reasoning_content` even when that turn was a normal text response. In a real multi-turn session, an earlier assistant answer without tool calls can still be present in history before a later tool call, so MiMo can continue returning `400 Param Incorrect` despite this patch.

```suggestion
  const assistant = message as ExtendedChatCompletionAssistantMessageParam;
  if (typeof assistant.reasoning_content === 'string') {
    return message;
  }

  return {
    ...assistant,
    reasoning_content: '',
  } as OpenAI.Chat.ChatCompletionMessageParam;
```

Please also update the `does not add reasoning_content to assistant turns without tool calls` test to expect the empty string, matching the DeepSeek provider behavior.

_— gpt-5.5 via Qwen Code /review_

**Review by wenshao — CHANGES_REQUESTED** [review]

File: packages/core/src/core/openaiContentGenerator/provider/mimo.ts (hunk 1 above)
Comment (DragonnZhang) [reply packages/core/src/core/openaiContentGenerator/provider/mimo.ts]: Addressed in b70a0516a: MiMo now returns OpenAIRequestContextOverrides from getRequestContextOverrides().

File: packages/core/src/core/openaiContentGenerator/pipeline.test.ts (hunk 2 above)
Comment (DragonnZhang) [reply packages/core/src/core/openaiContentGenerator/pipeline.test.ts:219]: Addressed in b70a0516a: added a conflicting precedence test where contentGeneratorConfig.splitToolMedia is true and the provider override is false.

File: packages/core/src/core/openaiContentGenerator/provider/mimo.ts (hunk 3 above)
Comment (DragonnZhang) [reply packages/core/src/core/openaiContentGenerator/provider/mimo.ts]: Addressed in b70a0516a: extracted shared reasoning_content replay logic into provider/utils.ts and wired both DeepSeek and MiMo through it.

File: packages/core/src/core/openaiContentGenerator/provider/mimo.ts (hunk 4 above)
Comment (DragonnZhang) [reply packages/core/src/core/openaiContentGenerator/provider/mimo.ts]: Addressed in b70a0516a: MiMo now injects empty reasoning_content on every assistant history turn, including non-tool-call turns, and the test expectation was updated.

File: packages/core/src/core/openaiContentGenerator/provider/mimo.test.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+describe('MiMoOpenAICompatibleProvider', () => {
+  it('is selected for Xiaomi MiMo hostnames', () => {
+    const provider = determineProvider(
+      createProviderConfig(),
+      createCliConfig(),
+    );
+
+    expect(provider).toBeInstanceOf(MiMoOpenAICompatibleProvider);
```
Comment (wenshao) [inline packages/core/src/core/openaiContentGenerator/provider/mimo.test.ts:39]: **[Suggestion]** Missing test for `determineProvider` ordering when both MiMo model name and OpenRouter hostname match.

A user configuring `baseUrl: 'https://openrouter.ai/api/v1'` + `model: 'mimo-v2.5-pro'` currently gets the MiMo provider (checked before OpenRouter in `determineProvider`). No test asserts this priority. If someone later reorders the checks, users routing MiMo models through OpenRouter would silently get the wrong provider — losing `reasoning_content` injection and `splitToolMedia: true`.

Consider adding:
```ts
it('is selected over OpenRouter when both hostname and model match', () => {
  const provider = determineProvider(
    createProviderConfig({
      baseUrl: 'https://openrouter.ai/v1',
      model: 'mimo-v2.5-pro',
    }),
    createCliConfig(),
  );
  expect(provider).toBeInstanceOf(MiMoOpenAICompatibleProvider);
});
```

_— qwen-latest-series-invite-beta-v34 via Qwen Code /review_

**Review by wenshao — APPROVED** [review]
No issues found. LGTM! ✅ _— qwen-latest-series-invite-beta-v34 via Qwen Code /review_

**Comment by wenshao** [comment]
## Maintainer test report — PR #4281

Built and validated locally in a dedicated `tmux` session (`pr4281`) on the `pr-4281` fetch branch (head `b70a051`, merge-base with `main` `33b2e0d`). Posting as a merge reference.

### Environment

- macOS 26.4.1 (arm64)
- Node v22.17.0 / npm 11.8.0
- Branch: `pull/4281/head` → local `pr-4281`
- Worktree: `/Users/wenshao/git/qwen-code-x6`
- Scope verified vs `main`: 9 files, +387/-27 (matches PR metadata)

### Results

| Stage | Result | Notes |
| --- | --- | --- |
| `npm install` | PASS | postinstall bundle ran cleanly |
| `npm run build` (workspaces) | PASS | only 1 unrelated lint warning in `vscode-ide-companion/editorGroupUtils.ts` (curly style) |
| `npm run bundle` (esbuild) | PASS | MiMo provider confirmed wired into `dist/chunks/chunk-*.js` (`MiMoOpenAICompatibleProvider`, `isMiMoProvider`, `getRequestContextOverrides`, `ensureReasoningContentOnAssistantMessage`, `xiaomimimo.com` all present) |
| Focused tests (`vitest run mimo.test.ts pipeline.test.ts` in `packages/core`) | PASS | **51/51** (mimo: 9, pipeline: 42) — author cited 50; current count is 51 |
| Full `packages/core` vitest | PASS for this PR | 8536 passed / 3 skipped / **1 pre-existing failure** unrelated to this change — reproduced on `origin/main` at `16f0fde`, see note below |
| `packages/core` typecheck | PASS | |
| Root `npm run typecheck` | **PASS** | The PR description flagged this as blocked by unrelated `packages/cli/src/serve/workspaceAgents.ts` / `workspaceMemory.ts` errors; those no longer reproduce on current `main`, so the PR is no longer gated by them |
| Lint on all 9 PR-touched files | PASS | zero issues |
| CLI smoke (`node dist/cli.js --version` / `--help`) | PASS | `0.15.11`, help renders |

### Pre-existing failure (NOT caused by this PR)

```
FAIL  src/core/anthropicContentGenerator/anthropicContentGenerator.test.ts
  > treats unset baseURL as Anthropic-native (SDK default targets api.anthropic.com)
AssertionError: expected 'claude-cli/1.2.3 (external, cli)' to contain 'QwenCode/1.2.3'
```

Reproduced on a clean worktree of `origin/main` (`16f0fde`) with the same test runner — same single-test failure, identical assertion. Not blocking for this PR; please track separately.

### Behavioral checks against the diff

- **Hostname matching** correctly rejects spoofed hosts: `https://api.xiaomimimo.com.evil.example/v1` → not selected (asserted by `mimo.test.ts`).
- **Model-name fallback** (`mimo-*` case-insensitive) selects MiMo even behind custom gateways — verified with `MiMo-V2.5-Pro` test.
- **`splitToolMedia` default** flips to `true` for MiMo via `getRequestContextOverrides`, and a user-provided `splitToolMedia: false` is still honored (provider override only fills when the user did not set it). Pipeline precedence test confirms provider override beats `contentGeneratorConfig.splitToolMedia`.
- **`reasoning_content` backfill** is now a shared util `ensureReasoningContentOnAssistantMessage` (extracted from `deepseek.ts`, reused by `mimo.ts`). Deepseek's existing behavior is preserved by the test suite. Existing non-empty `reasoning_content` is left untouched.

### Risk assessment

Low risk for merge.

- Provider selection is hostname-anchored AND model-prefix-fallback, with explicit spoof-rejection test.
- The shared-util extraction in `deepseek.ts` is a no-op refactor (covered by existing deepseek tests, still passing).
- The only intentional behavior change is `splitToolMedia` defaulting to `true` for MiMo, which is the documented fix and is opt-out via explicit `splitToolMedia: false`.

### Not covered locally

- Live MiMo API call (no credentials on this machine) — author noted the same gap.
- Windows / Linux / Docker / Podman / Seatbelt matrices — not run; matches PR's testing matrix.

### Reproduce

```bash
git fetch origin pull/4281/head:pr-4281 && git checkout pr-4281
npm install
cd packages/core && npx vitest run \
  src/core/openaiContentGenerator/provider/mimo.test.ts \
  src/core/openaiContentGenerator/pipeline.test.ts
cd ../.. && npm run typecheck
```

Recommendation: **safe to merge** once at least one reviewer signs off on the `splitToolMedia: true` default-flip semantics for MiMo. The pre-existing Anthropic UA test failure is independent and should be tracked separately.

**Review by pomelo-nwu — APPROVED** [review]
LGTM
